vcr/oidc4vci/wallet_client.go

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package oidc4vci
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+)
+
+type WalletAPIClient interface {
+	Metadata() OAuth2ClientMetadata
+	OfferCredential(ctx context.Context, offer CredentialOffer) error
+}
+
+var _ WalletAPIClient = (*defaultWalletAPIClient)(nil)
+
+// NewWalletAPIClient resolves the OAuth2 credential client metadata from the given URL.
+func NewWalletAPIClient(ctx context.Context, httpClient *http.Client, walletMetadataURL string) (WalletAPIClient, error) {
+	if walletMetadataURL == "" {
+		return nil, errors.New("empty wallet metadata URL")
+	}
+
+	metadata, err := loadOAuth2CredentialsClientMetadata(ctx, walletMetadataURL, httpClient)
+	if err != nil {
+		return nil, fmt.Errorf("unable to load OAuth2 credential client metadata (url=%s): %w", walletMetadataURL, err)
+	}
+
+	return &defaultWalletAPIClient{
+		httpClient: httpClient,
+		metadata:   *metadata,
+	}, nil
+}
+
+var _ WalletAPIClient = (*defaultWalletAPIClient)(nil)
+
+type defaultWalletAPIClient struct {
+	metadata   OAuth2ClientMetadata
+	httpClient *http.Client
+}
+
+func (c *defaultWalletAPIClient) Metadata() OAuth2ClientMetadata {
+	return c.metadata
+}
+
+func (c *defaultWalletAPIClient) OfferCredential(ctx context.Context, offer CredentialOffer) error {
+	offerJson, err := json.Marshal(offer)
+	if err != nil {
+		return err
+	}
+	requestURL := c.metadata.CredentialOfferEndpoint + "?credential_offer=" + url.QueryEscape(string(offerJson))
```

Comment:
path.Join?

Reply:
Not path joining, just specifying a query parameter. CredentialOfferEndpoint is a path without query params, according to the spec.